1 create initial processing chain

rosepearson opened the pull request:
A currently draft pull request where we can discuss the form of the initial tests associated with the package.

The package 'GeoFabrics' loads in data specified in an instruction.json file and runs a bare-bones processing chain to produce a water-tight DEM over the specified catchment. 

_Later issues will deal with ensuring that the produced DEM is hydrologically conditioned._

The ./src/main.py and ./instruction.json files currently act as examples for usage. These will be replaced by the test1 under root /tests/test1

_**A still unfulfilled requirement of the PR in my mind is replacement/utilisation of these by the ./tests/test1 initial test.**_ As part of this PR I would like to get input on the form of this test before creating synthetic data.

LukeParky commented on src/GeoFabrics/geometry.py:
```diff
@@ -0,0 +1,316 @@
+    
+    @property    
+    def land_and_foreshore(self):
+        """ Return the catchment and and foreshore region """
```
catchment and _land_ foreshore region

LukeParky commented on src/GeoFabrics/geometry.py:
```diff
@@ -0,0 +1,316 @@
+            
+        return self._lidar_extents
+    
+    def load_lidar_extents(self, lidar_extents_string): # expect to be filepath
```
You could do this type comment as a type hint such as in [type hint](https://docs.python.org/3/library/typing.html)  or other ways with in-built Python support

rosepearson replied:
Good idea. I've added for most of the class constructors as well

LukeParky commented on instruction.json:
```diff
@@ -0,0 +1,24 @@
+      "resolution": 10
+    },
+    "data_paths": {
+      "catchment_boundary": "C:\\Users\\pearsonra\\Documents\\data\\Wakanae\\Small_test_site\\initial_data\\boundary\\boundary.shp",
```
Could these be in a relative path format, such as in the test example?

rosepearson replied:
They could be - I was thinking of removing this from version control once the test is created and working. I will leave for now and remove prior to the end of this pr.

LukeParky reviewed (changes requested):
Looks like good structure, just some minor changes that I suggested

LukeParky commented on src/GeoFabrics/dem.py:
```diff
@@ -0,0 +1,146 @@
+        self.set_foreshore = set_foreshore
+        self._dem = rioxarray.rioxarray.open_rasterio(dem_file, masked=True)
+        
+        self.__set_up()
```
I'm unsure of the use of a double underscore here, compared to a single one. I'm not telling you to change it but just highlighting it

rosepearson replied:
I did some reading (link to docs below), which indicates that a double underscore should be used if you want to do name-mangling and a single underscore should be used for private variables - which is what I'm after so I'll change it thanks! 

https://docs.python.org/3/tutorial/classes.html#private-variables

rosepearson commented:
I'm planning to make a test dataset with the topography represented in the attached image. 

The datasets would be:
* LiDAR - 1 pt per metre over the middle 500x500m2 area as a LAZ file (starting 25m offshore)
* Reference DEM - a 15m resolution extending over the entire land region and extending 25m offshore
* Bathymetry contours - a 5m, 10m and 20m depth contour
 As well as the land and catchment bounding boxes in the image.

@AliceHarang, what do you think?

![image](https://user-images.githubusercontent.com/22883860/123579007-1dcb7b00-d82b-11eb-894b-6d926e2be7f1.png)

AliceHarang commented:
The topographic test dataset seems great ! Perhaps you can take a smaller region at the back behind the "top-range" line...
Do you think about adding some perturbations (noise or larger ones) in it?

rosepearson commented:
> The topographic test dataset seems great ! Perhaps you can take a smaller region at the back behind the "top-range" line...
> Do you think about adding some perturbations (noise or larger ones) in it?

Yes - I wondered that too. Maybe I'll just have the DEM extend just beyond the range-top.

I was also wondering about adding some noise - but thought I wouldn't for the very first test case (keep it very simple) and that could be done later as a whole new test. I could probably do it for this PR though.

LukeParky reviewed (changes requested):
Looks good to me. Resolve the conversation on `instruction.json` and it's approved

LukeParky commented on setup.py:
```diff
@@ -0,0 +1,54 @@
+        'Operating System :: OS Independent',
+    ],
+    python_requires='>=3.6',
+    # install_requires=['GeoFabrics'],
+    extras_require={
+        'dev': ['check-manifest'],
+        # 'test': ['coverage'],
+    },
+    # entry_points={
+    #     'console_scripts': [  # This can provide executable scripts
+    #         'run=examplepy:main',
+    # You can execute `run` in bash to run `main()` in src/examplepy/__init__.py
+    #     ],
+    # },
```
Remove commented code

rosepearson replied:
done!

LukeParky commented on src/GeoFabrics/processor.py:
```diff
@@ -0,0 +1,111 @@
+        z_label = self.instructions['instructions']['instructions']['bathymetry_contours_z_label'] if  \
+            'bathymetry_contours_z_label' in self.instructions['instructions']['instructions'] else None
+        self.bathy_contours = geometry.BathymetryContours(self.instructions['instructions']['data_paths']['bathymetry_contours'][0], self.catchment_geometry, z_label = z_label)
+        #bathy_points = geometry.BathymetryPoints(self.instructions['instructions']['data_paths']['bathymetry_points'][0], catchment_geometry)
```
remove commented code

LukeParky reviewed (changes requested):
Unit test isn't passing on my machine

[test_output.txt](https://github.com/rosepearson/GeoFabrics/files/6744125/test_output.txt)

LukeParky commented on README.md:
```diff
@@ -1,18 +1,23 @@
+with open(r'path\to\file\\', 'r') as file_pointer:
             instructions = json.load(file_pointer)
-runner = processor.GeoFabricsGenerator(instructions)
+runner = GeoFabrics.processor.GeoFabricsGenerator(instructions)
```
geofabrics, not GeoFabrics

rosepearson replied:
thanks - fixed

rosepearson commented:
Hi Luke,
I think I have addressed all of your suggestions, and the package is now ready to be tested.
Specifically, the testing I did is: 
1. run main.py with my own data and instruction file (I don't think you need to do this), 
2. run the unit tests and check they pass. 
3. import the package using pip install (see readme) into a different conda environment and check you can import geofabrics in Python.

rosepearson commented:
> Unit test isn't passing on my machine
> 
> [test_output.txt](https://github.com/rosepearson/GeoFabrics/files/6744125/test_output.txt)

Thanks! I've moved to making an almost equal comparison as discussed as the differences is in the order e-13 and only present in around 20% of the locations.
